### PR TITLE
subDirectory for Files object

### LINKED ISSUE #277: Sub-folder creation
--------
Issue Description
--------
##### What did you expect to see happen?
If I specify in the <Files> block a file that is contained in a sub-folder, I would expect RAVEN to 
recreate that sub-folder in the workingDir/StepName/1/ folder.

##### What did you see instead?
The file is just copied in the realization folder or crash

##### Do you have a suggested fix for the development team?
copy 

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [ ] 1. If the issue is a defect, is the defect fixed?
- [ ] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [ ] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [ ] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [ ] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address?
Closes  #277

##### What are the significant changes in functionality due to this change request?
Adding the functionality for locating Files in a path relative to the workingDir.

### CODE REVIEW

File: framework/Models/Code.py [hunk 1]
```diff
@@ -270,11 +270,15 @@ def initialize(self,runInfoDict,inputFiles,initDict=None):
     self.workingDir               = os.path.join(runInfoDict['WorkingDir'],runInfoDict['stepName']) #generate current working dir
     runInfoDict['TempWorkingDir'] = self.workingDir
     for inputFile in inputFiles:
-      shutil.copy(inputFile.getAbsFile(),self.workingDir)
+      subSubDirectory = os.path.join(self.workingDir,inputFile.subDirectory)
+      if inputFile.subDirectory.strip() != "" and not os.path.exists(subSubDirectory):
+        os.mkdir(subSubDirectory)
+      shutil.copy(inputFile.getAbsFile(),subSubDirectory)
     self.oriInputFiles = []
     for i in range(len(inputFiles)):
```
Comment (maljovec) [inline framework/Models/Code.py]: These loops (line 272-276 & 275-281) look very similar, is there a reason that they cannot be merged into one?

```python
    self.oriInputFiles = []
    for inputFile in inputFiles:
        subSubDirectory = os.path.join(self.workingDir,inputFile.subDirectory)
        if inputFile.subDirectory.strip() != "" and not os.path.exists(subSubDirectory):
            os.mkdir(subSubDirectory)
        shutil.copy(inputFile.getAbsFile(),subSubDirectory)
        self.oriInputFiles.append(copy.deepcopy(inputFile))
        self.oriInputFiles[-1].setPath(subSubDirectory)

```

File: doc/user_manual/files.tex [hunk 2]
```diff
@@ -29,6 +29,16 @@ \section{Files}
   directly use this attribute, but it is availabe in the CodeInterface.
   If not provided, defaults to True.
   %
+  
+  \item \xmlAttr{subDirectory}, \xmlDesc{optional string attribute}, sub-directory
+  that should be created in the perturbation process. If specified, the file will be
+  placed in the subdirectory. For example, in a \textit{MultiRun} step using a \textit{MonteCarlo}
```
Comment (maljovec) [inline doc/user_manual/files.tex]: Is the Sampler used relevant here? This seems like extra information not used in determining the path.

File: framework/Models/Code.py [hunk 3]
```diff
@@ -308,8 +312,11 @@ def createNewInput(self,currentInput,samplerType,**kwargs):
     if not os.path.exists(subDirectory):
       os.mkdir(subDirectory)
     for index in range(len(newInputSet)):
-      newInputSet[index].setPath(subDirectory)
-      shutil.copy(self.oriInputFiles[index].getAbsFile(),subDirectory)
+      subSubDirectory = os.path.join(subDirectory,newInputSet[index].subDirectory)
+      if newInputSet[index].subDirectory.strip() != "" and not os.path.exists(subSubDirectory):
+        os.mkdir(subSubDirectory)
```
Comment (maljovec) [inline framework/Models/Code.py]: Is it important to create the directory in two different places? I am just wondering if between `createNewInput` and `initialize`, one of these will always get to this line first? This would essentially make the other one into dead code. I don't know which is the "right" place and if there are use cases where both may occur.

File: framework/PostProcessors/BasicStatistics.py [hunk 4]
```diff
@@ -1085,16 +1085,18 @@ def run(self, inputIn):
       @ In,  inputIn, object, object contained the data to process. (inputToInternal output)
       @ Out, outputDict, dict, Dictionary containing the results
     """
-    input = self.inputToInternal(inputIn)
+    inputAdapted = self.inputToInternal(inputIn)
```
Comment (maljovec) [inline framework/PostProcessors/BasicStatistics.py]: Good catch. We should make sure this doesn't occur anywhere else in the code since this was overwriting the built-in [input](https://docs.python.org/3/library/functions.html#input) function.

File: framework/Samplers/DynamicEventTree.py [hunk 5]
```diff
@@ -300,7 +300,7 @@ def computeConditionalProbability(self,index=None):
       self.endInfo[index]['branchChangedParams'][key]['changedConditionalPb'] = []
       self.endInfo[index]['branchChangedParams'][key]['unchangedConditionalPb'] = parentCondPb*float(self.endInfo[index]['branchChangedParams'][key]['unchangedPb'])
       for pb in range(len(self.endInfo[index]['branchChangedParams'][key]['associatedProbability'])):
-        self.endInfo[index]['branchChangedParams'][key]['changedConditionalPb'].append(parentCondPb*float(self.endInfo[index]['branchChangedParams'][key]['associatedProbability'][pb]))
+        self.endInfo[index]['branchChangedParams'][key]['unchangedConditionalPb'].append(parentCondPb*float(self.endInfo[index]['branchChangedParams'][key]['associatedProbability'][pb]))
```
Comment (maljovec) [inline framework/Samplers/DynamicEventTree.py]: Seems unrelated. For the sake of posterity could you briefly explain for future generations why this variable was changed? Was it wrong to call it "changedConditionalPb" before or did the computation of this value change?

File: tests/framework/CodeInterfaceTests/test_generic_interface.xml [hunk 6]
```diff
@@ -27,6 +27,7 @@
     <Input name="one.xml" type="">one.xml</Input>
     <Input name="inp.two" type="">inp.two</Input>
     <Input name="mesh" type="">dummy.e</Input>
+    <Input name="a_dummy_file_for_subdirectory" type="" subDirectory="testSubDirectory">dummy_file_for_subdirectory.dummy</Input>
```
Comment (maljovec) [inline tests/framework/CodeInterfaceTests/test_generic_interface.xml]: How exactly does this work? Is this file generated or an input file? Or does it get copied from the working directory into some subdirectory?

File: framework/Samplers/DynamicEventTree.py (hunk 5 above)
Comment (alfoa) [inline framework/Samplers/DynamicEventTree.py]: sorry it is my mistake.... I revert it

File: doc/user_manual/files.tex (hunk 2 above)
Comment (alfoa) [inline doc/user_manual/files.tex]: removed

File: tests/framework/CodeInterfaceTests/test_generic_interface.xml (hunk 6 above)
Comment (alfoa) [inline tests/framework/CodeInterfaceTests/test_generic_interface.xml]: This is a UserGeneated input file. The attribute just specifies that a subDirectory needs to be generated in the sampling process. For example,
In a MultiRun, we will copy the file ```dummy_file_for_subdirectory.dummy``` from ```workingDir/subDirectory/``` to the location ```workingDir/stepName/1/subDirectory/```

File: framework/Models/Code.py (hunk 3 above)
Comment (alfoa) [inline framework/Models/Code.py]: The reason why we do that (for now) is the fact that just at the "createNewInput" stage we know the "prefix" of the realization and so we can create the folder "workingDir/stepName/%prefix%"

File: framework/Models/Code.py (hunk 1 above)
Comment (alfoa) [inline framework/Models/Code.py]: Done

**Comment by alfoa** [comment]
Addressed Dan's comments

File: framework/PostProcessors/BasicStatistics.py (hunk 4 above)
Comment (alfoa) [inline framework/PostProcessors/BasicStatistics.py]: yep

File: tests/framework/CodeInterfaceTests/test_generic_interface.xml (hunk 6 above)
Comment (maljovec) [inline tests/framework/CodeInterfaceTests/test_generic_interface.xml]: We have the option to delete these files, though right? If the file is read-only can we create a symbolic link to it? In case the file is large, it would allow us to bypass some overhead and save on storage. If we cannot guarantee the file is read-only, then all of this is irrelevant.

Comment (alfoa) [inline tests/framework/CodeInterfaceTests/test_generic_interface.xml]: The Inputs there are perturb-able  by the code interface. If we create a symbolic link we are going to perturb the original ones and we do not know what we are actually running.

Now if your suggestion is to create another option to specify if some files can be just "pointed" and not copied, I agree but it should be addressed in another PR/Issue

Comment (maljovec) [inline tests/framework/CodeInterfaceTests/test_generic_interface.xml]: Got it.

File: framework/Models/Code.py (hunk 3 above)
Comment (maljovec) [inline framework/Models/Code.py]: So, it looks like from the [code coverage](http://mooseframework.org/docs/PRs/285/python-coverage/_home_moosetest_civet_client_root_0_raven_framework_Models_Code_py.html) the directory creation in `initialize` is never executed. Do you anticipate this to change, or should we leave this in?

Comment (maljovec) [inline framework/Models/Code.py]: Actually, neither does the one from `createNewInput`. Did the amended test case not actually need to create any subdirectories?

**Comment by maljovec** [comment]
----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True. __Open Issue__
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync. __N/A__
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added? __N/A__

File: framework/Models/Code.py (hunk 3 above)
Comment (maljovec) [inline framework/Models/Code.py]: We should determine if these lines of code are reachable currently. If not, then it may make more sense to leave them as commented out or for now put a comment above them that states they may be valuable pending future development. I can try to amend this and merge it as this was my only outstanding issue.
